src/sources/ping/eventfd.rs

```diff
@@ -0,0 +1,190 @@
+//! Eventfd based implementation of the ping event source.
+//!
+//! # Implementation notes
+//!
+//! The eventfd is a much lighter signalling mechanism provided by the Linux
+//! kernel. Rather than write an arbitrary sequence of bytes, it only has a
+//! 64-bit counter.
+//!
+//! To avoid closing the eventfd early, we duplicate it in `make_ping()` with
+//! `dup()`. The returned fd can then be closed when all the senders are dropped
+//! without invalidating the source's fd.
+//!
+//! To differentiate between regular ping events and close ping events, we add 2
+//! to the counter for regular events and 1 for close events. In the source we
+//! can then check the LSB and if it's set, we know it was a close event. This
+//! only works if a close event never fires more than once.
+
+use std::{os::unix::io::RawFd, sync::Arc};
+
+use nix::sys::eventfd::{eventfd, EfdFlags};
+use nix::unistd::{close, dup, read, write};
+
+use super::PingError;
+use crate::{
+    generic::Generic, EventSource, Interest, Mode, Poll, PostAction, Readiness, Token, TokenFactory,
+};
+
+// These are not bitfields! They are increments to add to the eventfd counter.
+// Since the fd can only be closed once, we can effectively use the
+// INCREMENT_CLOSE value as a bitmask when checking.
+const INCREMENT_PING: u64 = 0x2;
+const INCREMENT_CLOSE: u64 = 0x1;
+
+#[inline]
+pub fn make_ping() -> std::io::Result<(Ping, PingSource)> {
+    let read = eventfd(0, EfdFlags::EFD_CLOEXEC | EfdFlags::EFD_NONBLOCK)?;
+    // We only have one fd for the eventfd. If the sending end closes it when
+    // all copies are dropped, the receiving end will be closed as well. We can
+    // avoid this by duplicating the FD.
+    let write = dup(read)?;
+
+    let source = PingSource {
+        event: Generic::new(read, Interest::READ, Mode::Level),
+    };
+
+    let ping = Ping {
+        event: Arc::new(FlagOnDrop { fd: write }),
+    };
+
+    Ok((ping, source))
+}
+
+// Helper functions for the event source IO.
+
+#[inline]
+fn send_ping(fd: RawFd, count: u64) -> std::io::Result<()> {
+    assert!(count > 0);
+    match write(fd, &count.to_ne_bytes()) {
+        // The write succeeded, the ping will wake up the loop.
+        Ok(_) => Ok(()),
+
+        // The counter hit its cap, which means previous calls to write() will
+        // wake up the loop.
+        Err(nix::errno::Errno::EAGAIN) => Ok(()),
+
+        // Anything else is a real error.
+        Err(e) => Err(e.into()),
+    }
+}
+
+#[inline]
+fn drain_ping(fd: RawFd) -> std::io::Result<u64> {
+    // The eventfd counter is effectively a u64.
+    const NBYTES: usize = std::mem::size_of::<u64>();
+    let mut buf = [0u8; NBYTES];
```

Comment:
Given this is documented to be a `u64` (and not an `usize`), the size is not platform-dependent, so that seems like a needlessly complex way to write `8` tbh (even though I guess the compile-time cost is negligible...)

Reply:
Zero, in fact because it's a `const fn` :wink:

Look, I am known to be pathological about this, but if I can find a way to avoid telling a computer a number I will. I will happily change it to 8 though because I recognise that I have a problem. (Do you object to the constant `NBYTES` itself or just the `size_of` call?)

Reply:
> Zero, in fact because it's a const fn

Hence why I was talking about *compile-time* cost. ^^

I don't have any strong feelings about this tbh, it just kinda surprised me to see this. No objection to keeping it like this if you strongly prefer it.

Reply:
Oh oops. Anyway, I am not married to it, I'll change it when I do @kchibisov's suggestion for the duplicated fd.